src/devices/bresser_leakage.c

```diff
@@ -0,0 +1,169 @@
+
+ */
+
+#include "decoder.h"
+
+static int bresser_leakage_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t const preamble_pattern[] = {0xaa, 0xaa, 0x2d, 0xd4};
+    data_t *data;
+    uint8_t msg[18];
+    uint32_t sensor_id;
+    uint8_t s_type;
+    uint8_t chan;
+    uint8_t startup;
+    uint8_t battery_ok;
+    int alarm;
+    int no_alarm;
+    int decode_ok;
+    uint8_t null_bytes;
+
+    if (bitbuffer->num_rows != 1
+            || bitbuffer->bits_per_row[0] < 160
+            || bitbuffer->bits_per_row[0] > 440) {
+        decoder_logf(decoder, 2, __func__, "bit_per_row %u out of range", bitbuffer->bits_per_row[0]);
+        return DECODE_ABORT_EARLY; // Unrecognized data
+    }
+
+    unsigned start_pos = bitbuffer_search(bitbuffer, 0, 0,
+            preamble_pattern, sizeof (preamble_pattern) * 8);
+
+    if (start_pos >= bitbuffer->bits_per_row[0]) {
+        return DECODE_ABORT_LENGTH;
+    }
+    start_pos += sizeof (preamble_pattern) * 8;
+
+    unsigned len = bitbuffer->bits_per_row[0] - start_pos;
+    if (len < sizeof(msg) * 8) {
+        decoder_logf(decoder, 2, __func__, "%u too short", len);
+        return DECODE_ABORT_LENGTH; // message too short
+    }
+
+    bitbuffer_extract_bytes(bitbuffer, 0, start_pos, msg, sizeof(msg) * 8);
+
+    decoder_log_bitrow(decoder, 2, __func__, msg, sizeof(msg) * 8, "");
+
+    // TODO: Find parity/checksum/checksum algorithm
+
+    sensor_id     = ((uint32_t)msg[2] << 24) | (msg[3] << 16) | (msg[4] << 8) | (msg[5]);
+    s_type        = msg[6] >> 4;
+    chan          = (msg[6] & 0x7);
+    battery_ok    = ((msg[7] & 0x30) != 0x00) ? 1 : 0;
+    startup       = (msg[6] >> 3) & 1;
+    alarm         = ((msg[7] & 0x80) == 0x80) ? 1 : 0;
+    no_alarm      = ((msg[7] & 0x40) == 0x40) ? 1 : 0;
+
+
+    null_bytes = msg[7] & 0xF;
+
+    for (int i=8; i<=15; i++) {
+        null_bytes |= msg[i];
+    }
+
+    // The parity/checksum/digest algorithm is currently unknown.
+    // We apply some heuristics to validate that the message is really from
+    // a water leakage sensor.
+    decode_ok = (s_type == SENSOR_TYPE_LEAKAGE) &&
+                (alarm != no_alarm) &&
+                (chan != 0) &&
+                (null_bytes == 0);
+
+   if (!decode_ok)
+       return 0;
+
+    /* clang-format off */
+    data = data_make(
+            "model",            "",             DATA_STRING, "Bresser-Leakage",
+            "id",               "",             DATA_FORMAT, "%08x",   DATA_INT,    sensor_id,
+            "channel",          "",             DATA_INT,    chan,
+            "battery_ok",       "Battery",      DATA_INT,    battery_ok,
+            "alarm",            "Alarm",        DATA_INT,    alarm,
+            "startup",          "Startup",      DATA_COND,   startup,  DATA_INT,    startup,
+            NULL);
+    /* clang-format on */
+
+    decoder_output_data(decoder, data);
+    return 1;
+}
+
+static char const *const output_fields[] = {
+    "model",
```

Comment:
Continued line indent is 8 spaces, not just 4. Also r_device below

Reply:
Fixed